File: proposed-projects/round-trip-experiment/reverse/20260909T012246Z_fbe56704/20260910T051241Z_fa3c6dd0/ported_implementation/src/grammar-graph/get-parent-stack-id.ts
import { Color } from './colorize.ts';
import type { Colorize } from './colorize.ts';
import type { GraphPointer } from './graph-pointer.ts';

export const getParentStackId = (pointer: GraphPointer, col: Colorize): string => {
  const stackIds: string[] = [];
  let parent = pointer.parent;
  while (parent) {
    const { stackId, pathId, stepId } = parent.node.meta;
    stackIds.push(`${stackId},${pathId},${stepId}`);
    parent = parent.parent;
  }
  return stackIds.map((id) => col(id, Color.RED)).join(col('<-', Color.GRAY));
};

File: packages/gbnf-experiment/data/20260909T134433Z_3f62598c/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from __future__ import annotations

from typing import TYPE_CHECKING, List, Optional

from .colorize import Color

if TYPE_CHECKING:  # pragma: no cover - typing only
    from .graph_pointer import GraphPointer


def get_parent_stack_id(pointer: "GraphPointer", col) -> str:
    stack_ids: List[str] = []
    parent: Optional["GraphPointer"] = pointer.parent
    while parent:
        meta = parent.node.meta
        stack_ids.append(f"{meta.stackId},{meta.pathId},{meta.stepId}")
        parent = parent.parent
    return col("<-", Color.GRAY).join(col(id_, Color.RED) for id_ in stack_ids)
